neurons/validators/src/services/miner_service.py

```diff
@@ -240,6 +251,88 @@ async def request_job_to_miner(self, payload: MinerJobRequestPayload):
                 exc_info=True,
             )
             return None
+        
+    def validate_digests(self, digests_in_list, duplicates):
+        # Check if any digest in digests_in_list is False
+        if any(not is_in_list for is_in_list in digests_in_list.values()):
+            return False
+
+        if duplicates:
+            return False
+
+        return True
+    
+    def get_docker_hub_digests(self, repositories):
+        """Retrieve all tags and their corresponding digests from Docker Hub."""
+        all_digests = {}  # Initialize a dictionary to store all tag-digest pairs
+
+        for repository in repositories:
+            try:
+                # Get authorization token
+                token_response = requests.get(
+                    f"https://auth.docker.io/token?service=registry.docker.io&scope=repository:{repository}:pull"
+                )
+                token_response.raise_for_status()  # Raise an error for bad responses
+                token = token_response.json().get("token")
+
+                # Find all tags
+                tags_response = requests.get(
+                    f"https://index.docker.io/v2/{repository}/tags/list",
+                    headers={"Authorization": f"Bearer {token}"}
+                )
+                tags_response.raise_for_status()
+                all_tags = tags_response.json().get("tags", [])
+
+                # Dictionary to store tag-digest pairs for the current repository
+                tag_digests = {}
+                for tag in all_tags:
+                    # Get image digest
+                    manifest_response = requests.head(
+                        f"https://index.docker.io/v2/{repository}/manifests/{tag}",
+                        headers={
+                            "Authorization": f"Bearer {token}",
+                            "Accept": "application/vnd.docker.distribution.manifest.v2+json"
+                        }
+                    )
+                    manifest_response.raise_for_status()
+                    digest = manifest_response.headers.get("Docker-Content-Digest")
+                    tag_digests[f"{repository}:{tag}"] = digest
+
+                # Update the all_digests dictionary with the current repository's tag-digest pairs
+                all_digests.update(tag_digests)
+
+            except requests.exceptions.RequestException as e:
+                print(f"Error retrieving data for {repository}: {e}")
+
+        return all_digests 
+
+    def check_digests(self, results, list_digests):
```

Comment:
these functions check_digests, check_duplicated_digests, validate_digests should be implemented on TaskService, not MinerService

Reply:
You can get the DockerHub Digests, before run the job and pass the result to TaskService

Fetching the DockerHub Digests could be done on MinerService or core/validator.py for better performance

Reply:
okay.